### PR TITLE
Feature/add changelog modal/178

### PR DESCRIPTION
**Pull request by GaetanGrd**
## Scope

[closes TICKET-178](https://github.com/Zagrios/bs-manager/issues/178)

## Implementation
When BSM is opened, a changelog is displayed if it has been updated. 
It can also be opened by clicking on the version number at the bottom of the settings page. 

## How to Test

- Launch and close BSM to adding var "haveBeenUpdated" in config files
- Set "haveBeenUpdated" in config file to "True"
- Launch BSM

Second test :
- Launch BSM and go settings pages
- Click on the version number.
## Emoji Guide

**For reviewers: Emojis can be added to comments to call out blocking versus non-blocking feedback.**

E.g: Praise, minor suggestions, or clarifying questions that don’t block merging the PR.

> 🟢 Nice refactor!

> 🟡 Why was the default value removed?

E.g: Blocking feedback must be addressed before merging.

> 🔴 This change will break something important

| | | |
| --- | --- | --- |
| Blocking | 🔴 ❌ 🚨 | RED |
| Non-blocking | 🟡 💡 🤔 💭 | Yellow, thinking, etc |
| Praise | 🟢 💚 😍 👍 🙌 | Green, hearts, positive emojis, etc |

### CODE REVIEW

File: src/shared/models/bs-launch/launch-changelog.interface.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export interface Changelog {
+  Title: string,
+  Body: string,
```
Comment (Zagrios) [inline src/shared/models/bs-launch/launch-changelog.interface.ts:3]: remove maj here

File: src/renderer/windows/App.tsx [hunk 2]
```diff
@@ -38,12 +39,16 @@ export default function App() {
     const location = useLocation();
     const navigate = useNavigate();
 
+  const updaterService = AutoUpdaterService.getInstance();
```
Comment (Zagrios) [inline src/renderer/windows/App.tsx:42]: Cannot use the useService hook here ?

File: src/renderer/services/auto-updater.service.ts [hunk 3]
```diff
@@ -44,6 +58,34 @@ export class AutoUpdaterService{
...
+        .then(data => resolve(data))
+        .catch(error => {
+          resolve(null);
+        });
+    });
+  }
+
+  public async openChangelog(force?: boolean): Promise<void> {
+    const isUpdated = await lastValueFrom(this.getHaveBeenUpdated());
+    if (!isUpdated && !force) {console.log("coucou"); return ;}
```
Comment (Zagrios) [inline src/renderer/services/auto-updater.service.ts:84]: remove console log

File: src/renderer/services/auto-updater.service.ts [hunk 4]
```diff
@@ -44,6 +58,34 @@ export class AutoUpdaterService{
...
+
+    const data = await this.getChangelogs();
+    if (!data) {console.log("coucou2"); return ;}
```
Comment (Zagrios) [inline src/renderer/services/auto-updater.service.ts:87]: remove console log

File: src/renderer/pages/settings-page.component.tsx [hunk 5]
```diff
@@ -70,6 +71,7 @@ export function SettingsPage() {
   const [appVersion, setAppVersion] = useState("");
   const nav = useNavigate();
   const t = useTranslation();
+  const updaterService = AutoUpdaterService.getInstance();
```
Comment (Zagrios) [inline src/renderer/pages/settings-page.component.tsx:74]: Cannot use the useService hook here ?

File: src/renderer/components/svgs/icons/search-icon.component.tsx [hunk 6]
```diff
@@ -2,7 +2,7 @@ import { CSSProperties } from "react";
 
 export function SearchIcon(props: {className?: string, style?: CSSProperties}) {
   return (
-    <svg className={props.className} style={props.style} xmlns="http://www.w3.org/2000/svg" viewBox="0 0 40 40" height="40" width="40">
+    <svg className={props.className} style={props.style} viewBox="0 0 16 16">
```
Comment (Zagrios) [inline src/renderer/components/svgs/icons/search-icon.component.tsx:5]: why these changes ?

File: src/renderer/components/modal/modal-types/changelog/changelog-modal.component.tsx [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+import { ModalComponent } from "renderer/services/modale.service";
+import BeatRunning from '../../../../../../assets/images/apngs/beat-running.png';
+import "./changelog-modal.component.css"
+import { BsmButton } from "renderer/components/shared/bsm-button.component";
+import { LinkOpenerService } from "renderer/services/link-opener.service";
+import { Changelog } from '../../../../../shared/models/bs-launch/launch-changelog.interface';
+import DOMPurify from "dompurify";
+
+export const ChangelogModal: ModalComponent<void, Changelog> = ({ resolver, data }) => {
+  const linkOpener: LinkOpenerService = LinkOpenerService.getInstance();
```
Comment (Zagrios) [inline src/renderer/components/modal/modal-types/changelog/changelog-modal.component.tsx:11]: Cannot use the useService hook here ?

File: src/main/ipcs/window-manager-ipcs.ts [hunk 8]
```diff
@@ -15,13 +18,15 @@ ipcMain.on("open-window-then-close-all", async (event, request: IpcRequest<AppWi
...
 
 ipcMain.on("close-all-windows", async (event, request: IpcRequest<AppWindow>) => {
+  if (request.args.includes("index.html")) {configService.set(HAVE_BEEN_UPDATED_KEY, false);}
     await launcher.restoreSteamVR();
     const windowManager = WindowManagerService.getInstance();
     windowManager.closeAllWindows(request.args);
 });
 
 ipcMain.on("close-windows", async (event, request: IpcRequest<AppWindow[]>) => {
+  if (request.args.includes("index.html")) {configService.set(HAVE_BEEN_UPDATED_KEY, false);}
```
Comment (Zagrios) [inline src/main/ipcs/window-manager-ipcs.ts:28]: wrong indentation here ?

File: src/main/ipcs/window-manager-ipcs.ts [hunk 9]
```diff
@@ -15,13 +18,15 @@ ipcMain.on("open-window-then-close-all", async (event, request: IpcRequest<AppWi
 });
 
 ipcMain.on("close-all-windows", async (event, request: IpcRequest<AppWindow>) => {
+  if (request.args.includes("index.html")) {configService.set(HAVE_BEEN_UPDATED_KEY, false);}
```
Comment (Zagrios) [inline src/main/ipcs/window-manager-ipcs.ts:21]: wrong indentation here ?

File: assets/jsons/changelogs/fr.json [hunk 10]
```diff
@@ -0,0 +1,5 @@
+{
+  "Title": "test DOMPurify <div></div>",
+  "Body":"<div><h1>Test de DOMPurify</h1><p>Ceci est un paragraphe de texte normal.</p><a href=\"https://example.com\">Ceci est un lien</a><script>alert(\"Ceci est un script dangereux\");</script><button onclick=\"alert('Ne pas cliquer!')\">Ne pas cliquer</button><iframe src=\"javascript:alert('xss')\"></iframe><img src=\"javascript:alert('xss');\"><a href=\"javascript:alert('xss');\">Lien dangereux</a><style>body {background-color: red;}</style><form action=\"https://example.com\"><input type=\"text\" name=\"username\"><input type=\"password\" name=\"password\"></form><object data=\"data:text/html;base64,PHNjcmlwdD5hbGVydCgnWFNTJyk8L3NjcmlwdD4=\"></object><!-- Ceci est un commentaire --></div>\n",
```
Comment (Zagrios) [inline assets/jsons/changelogs/fr.json:3]: change keys with the new interface without uppercases

empty the values

File: assets/jsons/changelogs/es.json [hunk 11]
```diff
@@ -0,0 +1,5 @@
+{
+  "Title": "es",
+  "Body":"es",
```
Comment (Zagrios) [inline assets/jsons/changelogs/es.json:3]: change keys with the new interface without uppercases

empty the values

File: assets/jsons/changelogs/en.json [hunk 12]
```diff
@@ -0,0 +1,5 @@
+{
+  "Title": "en",
+  "Body":"en",
```
Comment (Zagrios) [inline assets/jsons/changelogs/en.json:3]: change keys with the new interface without uppercases

empty the values

File: assets/jsons/changelogs/de.json [hunk 13]
```diff
@@ -0,0 +1,5 @@
+{
+  "Title": "de",
+  "Body":"de",
```
Comment (Zagrios) [inline assets/jsons/changelogs/de.json:3]: change keys with the new interface without uppercases

empty the values

File: src/renderer/services/auto-updater.service.ts [hunk 14]
```diff
@@ -44,6 +58,34 @@ export class AutoUpdaterService{
...
+
+  public async getChangelogs(): Promise<Changelog | null> {
+    const path = `https://raw.githubusercontent.com/GaetanGrd/bs-manager/feature/add-changelog-modal/178/assets/jsons/changelogs/${this.i18nService.currentLanguage.split("-")[0]}.json`
+    return new Promise((resolve) => {
+      fetch(path)
+        .then(response => {
+          if (!response.ok) {
+            throw new Error('Network response was not ok');
+          }
+          return response.json();
```
Comment (Zagrios) [inline src/renderer/services/auto-updater.service.ts:73]: add check to return null of the body of changlog is empty

File: src/shared/models/bs-launch/launch-changelog.interface.ts (hunk 1 above)
Comment (GaetanGrd) [reply src/shared/models/bs-launch/launch-changelog.interface.ts:3]: idem for interface name ?

File: src/renderer/pages/settings-page.component.tsx (hunk 5 above)
Comment (GaetanGrd) [reply src/renderer/pages/settings-page.component.tsx:75]: When I tried to implement it I ended up with lots of errors, so I thought I'd do it with the expected refacto...

File: src/renderer/components/svgs/icons/search-icon.component.tsx (hunk 6 above)
Comment (GaetanGrd) [reply src/renderer/components/svgs/icons/search-icon.component.tsx:5]: idk... wtf ?

File: src/main/ipcs/window-manager-ipcs.ts [hunk 15]
```diff
@@ -15,13 +18,15 @@ ipcMain.on("open-window-then-close-all", async (event, request: IpcRequest<AppWi
 });
 
 ipcMain.on("close-all-windows", async (event, request: IpcRequest<AppWindow>) => {
+    if (request.args.includes("index.html")) {configService.set(HAVE_BEEN_UPDATED_KEY, false);}
```
Comment (Zagrios) [inline src/main/ipcs/window-manager-ipcs.ts:21]: put haveBeenUpdated setter in the service

File: src/renderer/components/modal/modal-types/changelog/changelog-modal.component.tsx [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+import { BsmButton } from "renderer/components/shared/bsm-button.component";
+import { LinkOpenerService } from "renderer/services/link-opener.service";
+import { Changelog } from '../../../../../shared/models/bs-launch/launch-changelog.interface';
+import DOMPurify from "dompurify";
+import { useService } from '../../../../hooks/use-service.hook';
+import { OsDiagnosticService } from '../../../../services/os-diagnostic.service';
+
+export const ChangelogModal: ModalComponent<void, Changelog> = ({ resolver, data }) => {
+
+  useService(OsDiagnosticService);
```
Comment (Zagrios) [inline src/renderer/components/modal/modal-types/changelog/changelog-modal.component.tsx:14]: why do you need that ?

File: src/renderer/services/auto-updater.service.ts [hunk 17]
```diff
@@ -44,6 +56,34 @@ export class AutoUpdaterService{
         this.ipcService.sendLazy("install-update");
     }
 
+    public getHaveBeenUpdated() : Observable<boolean>{
+        return this.ipcService.sendV2<boolean>("have-been-updated");
+    }
+
+  public async getChangelogs(): Promise<Changelog | null> {
```
Comment (Zagrios) [inline src/renderer/services/auto-updater.service.ts:63]: if you remove "| null" you have an error ?

File: src/renderer/services/auto-updater.service.ts [hunk 18]
```diff
@@ -44,6 +56,34 @@ export class AutoUpdaterService{
...
+      }
+
+      const data = await response.json();
+      if (!data?.body) {
+        return;
+      }
+      return data;
+
+    } catch (error) {
+      return;
```
Comment (Zagrios) [inline src/renderer/services/auto-updater.service.ts:79]: no need return here, you can use the thing at line 15 in index.tsx to log the error

File: src/renderer/services/auto-updater.service.ts [hunk 19]
```diff
@@ -44,6 +56,34 @@ export class AutoUpdaterService{
...
+    }
+  }
+
+  public async openChangelog(): Promise<void> {
+    const data = await this.getChangelogs();
+    if (!data) {return ;}
```
Comment (Zagrios) [inline src/renderer/services/auto-updater.service.ts:85]: space between return and ";"

File: src/renderer/pages/settings-page.component.tsx [hunk 20]
```diff
@@ -161,6 +167,8 @@ export function SettingsPage() {
 
   const openLogs = () => ipcService.sendLazy("open-logs");
 
+  const openChangelog = () => setModalChangelogResponse(updaterService.openChangelog())
```
Comment (Zagrios) [inline src/renderer/pages/settings-page.component.tsx:170]: you store the result of openChangelog but the function always return a truthy value due to the promise

File: src/renderer/windows/App.tsx [hunk 21]
```diff
@@ -34,16 +35,20 @@ export default function App() {
...
 
+
+
     useEffect(() => {
         themeService.theme$.subscribe(() => {
             if(themeService.isDark || (themeService.isOS && window.matchMedia('(prefers-color-scheme: dark)').matches)){ return document.documentElement.classList.add('dark'); }
             document.documentElement.classList.remove('dark');
         });
 
+        lastValueFrom(updaterService.getHaveBeenUpdated()).then(v => v && updaterService.openChangelog());
```
Comment (Zagrios) [inline src/renderer/windows/App.tsx:51]: rename "v" to "haveBeenUpdated" for more clarity

**Review by Zagrios — CHANGES_REQUESTED** [review]

File: src/renderer/components/modal/modal-types/changelog/changelog-modal.component.tsx (hunk 16 above)
Comment (GaetanGrd) [reply src/renderer/components/modal/modal-types/changelog/changelog-modal.component.tsx:14]: useService(OsDiagnosticService); ?

**Comment by Zagrios** [comment]
😭

**Comment by Zagrios** [comment]
😭😭

**Comment by GaetanGrd** [comment]
Just to let you know, I was on holiday, so I wasn't able to make any progress on this review, :) I'll be back soon!

File: src/renderer/services/auto-updater.service.ts (hunk 17 above)
Comment (GaetanGrd) [reply src/renderer/services/auto-updater.service.ts:63]: no

File: src/renderer/services/auto-updater.service.ts (hunk 18 above)
Comment (GaetanGrd) [reply src/renderer/services/auto-updater.service.ts:79]: it's realy useless ? 
if i remove the return, have a error "Not all code paths return a value." 👀

Comment (GaetanGrd) [reply src/renderer/services/auto-updater.service.ts:82]: catch (error) {
      this.ipcService.sendLazy("log-error", { args: error });
    }

Comment (GaetanGrd) [reply src/renderer/services/auto-updater.service.ts:82]:
```suggestion
  } catch (error) {
      this.ipcService.sendLazy("log-error", { args: error });
    }
```

**Comment by Zagrios** [comment]
😭

**Comment by GaetanGrd** [comment]
don't cry
i play beat saber 👀

**Comment by GaetanGrd** [comment]
😭
